back-end/apis/model.py

```diff
@@ -139,160 +138,130 @@ def UploadModel():
-            )
-        except Exception as e:
-            traceback.print_exc()
-            clear_model_temp_files(base_dir, saving_id)
-            return RResponse.abort(400, f"Failed to initialize the custom model. {e}")
-    elif "pretrained" in metadata:  # If the model is predefined
-        pretrained = bool(int(metadata.get("pretrained")))
-        num_classes = int(metadata.get("num_classes"))
-        try:
-            model = init_predefined_model(class_name, pretrained, num_classes)
-            with open(code_path, "w") as code_file:
-                code_file.write(f"num_classes = {num_classes}")
-        except Exception as e:
-            traceback.print_exc()
-            return RResponse.abort(
-                400, f"Failed to initialize the predefined model. {e}"
-            )
-    else:
-        return RResponse.abort(400, "The model is neither custom nor predefined.")
+        code_path = os.path.join(
+            RServer.get_server().base_dir,
+            "generated",
+            "models",
+            "code",
+            f"{saving_id}.py",
+        )
+        weight_path = os.path.join(
+            RServer.get_server().base_dir,
+            "generated",
+            "models",
+            "ckpt",
+            f"{saving_id}.pth",
+        )
 
-    # Get the weight file and save it to a temporary location if it exists
-    if "weight_file" in request.files:
-        weight_file = request.files.get("weight_file")
-        try:
-            weight_path = os.path.join(
-                RServer.get_server().base_dir, "generated", "models", f"{saving_id}.pth"
+        # Get the model's class name
+        class_name = metadata.get("class_name")
+
+        predefined = bool(int(metadata.get("predefined")))
+
+        # Save the model's code definition and initialize the model
+        if not predefined:  # If the model is custom
+            # Get the model definition code and save it to a temporary file
+            code = request.form.get("code")
+            save_code(code, code_path)
+            # Initialize the model
+            try:
+                model = RModelWrapper.init_custom_model(
+                    code_path, class_name, RServer.get_model_wrapper().device
+                )
+            except Exception as e:
+                traceback.print_exc()
+                clear_model_temp_files(code_path, weight_path)
+                return RResponse.fail(
+                    f"Failed to initialize the custom model. {e}", 400
+                )
+        elif predefined:  # If the model is predefined
+            pretrained = bool(int(metadata.get("pretrained")))
+            num_classes = int(metadata.get("num_classes"))
+            # TODO: Stop relying on the user to provide the number of classes
+            code = f"num_classes = {num_classes}"
+            save_code(code, code_path)
+            try:
+                model = RModelWrapper.init_predefined_model(
```

Comment:
When we load a model, we will always check model weight path and:
- If weight path is empty, random initialize
- If weight path exists, load it

If a model is pre-defined and pre-trained, when we load, we also want to load the weights. The question is, how do we know a model is pre-trained or not when pulling it out from DB? There is no `pre-trained` field is not in the metadata.

Two ways:
- For pre-defined pre-trained models, we might also need to save model weights. This adds logical consistency, but sacrifices storage (model weight stored twice, once in pip cache, once in robustar).
- Add another field `pre-trained` in metadata, and add an additional check for this field when loading. I personally prefer this. This is also a good information to track from user's perspective.

Reply:
Ok, I will add `pretrained` as a new field.